### ISSUE TITLE
Unify naming methods into a single method + options builder type

### ISSUE THREAD

**Issue by fitzgen** [servo/rust-bindgen#201]
Right now we have a few different methods for getting something name, plus a few different parameters controlling what kind of name you end up getting from it. I think it would be a nice clean up if we had only a single method for getting a name, that took an options type that described what kind of name we wanted. We could make the naming options a builder class. The defaults would be sanem

Imagining something like this:

```rust
// Create a name options builder.
item.name()
    // Use namespaces, regardless what the BindgenOptions
    // were configured with.
    .with_namespaces()
    // Do not use mangle the name to avoid Rust keywords.
    .no_rust_mangling()
    // Do not automatically deref this item into its replacement.
    .ignore_replacements()
    // Finish the builder and compute the actual name.
    .get(context)
```

Details aren't important. What is important is:

* Sane defaults based on how BindgenOptions was configured (for example, use whatever option was specified for namespaces by default)

* Force individually name options on/off regardless whether BindgenOptions were configured to have that option on/off

* Control over whether replacements are taken into account

**Comment by fitzgen** [comment]
And we would probably want this single unified naming method to be part of the `ItemCanonicalName` trait so everything can use it (and maybe drop `Item` to be just `CanonicalName`).

**Comment by impowski** [comment]
I can take it. And it would be good if you give me a little guide where to start.

**Comment by fitzgen** [comment]
@impowski, great!

Since I wrote this issue up, the situation has improved a little bit. We've separated getting a name within a namespace from getting a full name including namespaces and potentially namespace mangling.

Right now the naming variations we support are all permutations of

* with or without namespace mangling. eg C++'s `foo::bar::Baz` becomes `foo_bar_Baz` (ie the `within_namespace` parameter to `Item::real_canonical_name`)

* for replacement checking vs for code generation (ie the `for_name_checking` parameter to `Item::real_canonical_name`)

I would hold off on any hypothetical other naming options until we find a use for them (I have some ideas, but we shouldn't bog ourselves down yet).

So, I suggest we start with something like this:

* create a `NameOptions` builder struct with flags for the options above

* make `Item::real_canonical_name` et al take the options struct as a parameter instead of multiple bool parameters

* add convenience methods so we can have usage kind of like the example in the original issue comment

    * on `Item`, define `fn name(&self) -> NameOptions`

    * on `NameOptions`, define `fn get(&BindgenContext) -> String` which calls `Item::real_canonical_name`

I hope that's enough to get you started, if anything is unclear, don't hesitate to reach out to me :)
